### PR TITLE
Check all filterSubjects of consumers when purging a stream

### PR DESCRIPTION
**Pull request by svenfoo**
Closes https://github.com/nats-io/nats-server/issues/4826

### CODE REVIEW

File: server/consumer.go [hunk 1]
```diff
@@ -3374,6 +3374,25 @@ func (o *consumer) isFilteredMatch(subj string) bool {
 	return false
 }
 
+// Check if the candidate filter subject is equal to or a subset match
```
Comment (derekcollison) [inline server/consumer.go:3377]: Candidate is not a filter subject per se, and needs to be a literal.

File: server/consumer.go [hunk 2]
```diff
@@ -3374,6 +3374,25 @@ func (o *consumer) isFilteredMatch(subj string) bool {
...
+// of one of the filter subjects.
+// Lock should be held.
+func (o *consumer) isEqualOrSubsetMatch(subj string) bool {
+	for _, filter := range o.subjf {
```
Comment (derekcollison) [inline server/consumer.go:3381]: @Jarema do we keep track if the filter has a wildcard?

Comment (Jarema) [reply server/consumer.go:3381]: yes, the `subjectFilter` struct has bool field `hasWildcard`.

File: server/consumer.go (hunk 1 above)
Comment (svenfoo) [reply server/consumer.go:3377]: Oh, does it? Maybe I didn't fully understand how this code is being used then. I was assuming that it gets called when `nats stream purge --subject=SUBJECT` is used. And I think I have successfully used this command with subjects that have a wildcard.

File: server/consumer.go (hunk 2 above)
Comment (derekcollison) [reply server/consumer.go:3381]: ok, then we can skip comparison in this loop if so? Might not have any consequences, but strict equality with wildcards looks strange to me so would rather avoid here if possible.

Comment (svenfoo) [reply server/consumer.go:3381]: I've pushed a commit that implements this. It does not seem to have an effect on the tests though. There are two tests in the set where the tested and the matched subjects are equal and have a wildcard.  For both cases the result is still `true`, albeit it's probably more computationally complex to come to that decision.

If we have a consumer that filters for `foo.>` and purge the stream with the filter `foo.>`, then the consumer's sequence needs to be updated, right? Same if the consumer filters for `foo.*.bar` and we purge with `foo.*.bar`. It seems to be correct to check for equality here, even if the filter has a wildcard. But maybe I am missing cases where this doesn't hold true?

**Review by derekcollison — APPROVED** [review]
LGTM - Thanks!

**Pull request merged**